libraries/AP_BattMonitor/AP_BattMonitor_FuelLevel_Reventech.h

```diff
@@ -0,0 +1,43 @@
+#pragma once
+
+#include "AP_BattMonitor_Backend.h"
+
+#if AP_BATTERY_FUELLEVEL_REVENTECH_ENABLED
+
+#include <Filter/Filter.h>                     // Filter library
+#include <Filter/AverageFilter.h>      // AverageFilter class (inherits from Filter class)
+#include "AP_BattMonitor.h"
+
+#define FILTER_MAX 60
+
+class AP_BattMonitor_FuelLevel_Reventech : public AP_BattMonitor_Backend
+{
+public:
+
+    /// Constructor
+    AP_BattMonitor_FuelLevel_Reventech(AP_BattMonitor &mon, AP_BattMonitor::BattMonitor_State &mon_state, AP_BattMonitor_Params &params);
+
+    static const struct AP_Param::GroupInfo var_info[];
+
+    /// Read the battery voltage and current.  Should be called at 10hz
+    void read() override;
+
+    /// returns true if battery monitor provides consumed energy info
+    bool has_consumed_energy() const override { return true; }
+
+    /// returns true if battery monitor provides current info
+    bool has_current() const override { return true; }
+
+    void init(void) override {}
+
+private:
+
+    AP_Float _ls800_max_fuel_level_litres;
+    AP_Int8  _pin;
+
+    AP_HAL::AnalogSource *_analog_source;
+
+    AverageFilter<float,float,FILTER_MAX> _ls800_filter;
```

Comment:
FILTER_MAX is 60, so averaging over 60 samples. How fast do samples come in?

Reply:
I would like to have the battery monitor on AP_Periph update at 1Hz instead of 10Hz so having 60 samples would result in averaging over 1 minute. If kept at 10Hz, I would require 600 samples which is not allowed by the average_filter object as it takes a maximum of an uint8_t which is 255, unless I implement my own averaging function, which I can do if preferred.

Reply:
We can't afford to slow down all the battery monitor polling, as you need the other ones to still update at 10Hz. You could look at limiting how often just this backend actually updates, but otherwise I'd prefer to see us just implement a different filter here. AverageFilter accumulates a buffer of every entry. So you are using 4 bytes per, this filter needs 240 bytes of RAM at the moment which seems very excessive. Why not just use a low pass filter that doesn't have to retain a very large buffer?

Reply:
Hi @WickedShell ok, thanks for clarifying the buffer limitations of average filter. Will give the Low Pass Filter a try and remove the average filter lines.

Reply:
outdated